File: Trabajo1_LeyvaBanda/Tarea10 (BucketSort)/Tarea10.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void insertionSort(vector<float>& bukt) {
    for (int j = 1; j < bukt.size(); j++) {
        float val = bukt[j];
        int k = j - 1;
        while (k >= 0 && bukt[k] > val) {
            bukt[k + 1] = bukt[k];
            k--;
        }
        bukt[k + 1] = val;
    }
}

void bucketSort(vector<float>& arr) {
    int s = arr.size();
    vector<vector<float>> buckets(s);

    for (float j : arr) {
        int bi = s * j;
        if (bi >= s) bi = s - 1;
        buckets[bi].push_back(j);
    }

    for (auto& bukt : buckets)
        insertionSort(bukt);

    int idx = 0;
    for (auto& bukt : buckets)
        for (float j : bukt)
            arr[idx++] = j;
}

int main() {
    vector<float> arr = {0.77, 0.16, 0.38, 0.25, 0.71, 0.93, 0.22, 0.11, 0.24, 0.67};
    cout << "Arreglo antes de ordenar:\n";
    for (float x : arr) cout << x << " ";
    cout << endl;

    bucketSort(arr);

    cout << "Arreglo después de ordenar:\n";
    for (float x : arr) cout << x << " ";
    cout << endl;
}

File: Trabajo1_LeyvaBanda/Tarea10 (BucketSort)/Tarea10.js
function insertionSort(bukt) {
  for (let j = 1; j < bukt.length; j++) {
    let val = bukt[j];
    let k = j - 1;
    while (k >= 0 && bukt[k] > val) {
      bukt[k + 1] = bukt[k];
      k--;
    }
    bukt[k + 1] = val;
  }
}

function bucketSort(arr) {
  let s = arr.length;
  let buckets = Array.from({ length: s }, () => []);

  for (let j of arr) {
    let bi = Math.floor(s * j);
    if (bi >= s) bi = s - 1;
    buckets[bi].push(j);
  }

  for (let bukt of buckets)
    insertionSort(bukt);

  let idx = 0;
  for (let bukt of buckets)
    for (let j of bukt)
      arr[idx++] = j;
}

let arr = [0.77, 0.16, 0.38, 0.25, 0.71, 0.93, 0.22, 0.11, 0.24, 0.67];
console.log("Arreglo antes de ordenar:");
console.log(arr.join(" "));

bucketSort(arr);

console.log("Arreglo después de ordenar:");
console.log(arr.join(" "));
